internal/output/html.go

```diff
@@ -0,0 +1,494 @@
+		sourceName := packageSource.Source.String()
+		if strings.Contains(sourceName, "/usr/lib/") {
+			continue
+		}
+
+		// Process vulnerabilities for each source
+		sourceResult := processSource(packageSource)
+		if sourceResult == nil {
+			continue
+		}
+
+		sourceList, ok := ecosystemMap[sourceResult.Ecosystem]
+		if ok {
+			ecosystemMap[sourceResult.Ecosystem] = append(sourceList, *sourceResult)
+		} else {
+			ecosystemMap[sourceResult.Ecosystem] = []HTMLSourceResult{*sourceResult}
+		}
+
+		updateCount(&resultCount, &sourceResult.HTMLVulnCount)
+	}
+
+	// Build the final result
+	return buildHTMLResult(ecosystemMap, resultCount)
+}
+
+// processSource processes a single source (lockfile or artifact) and returns an SourceResult.
+func processSource(packageSource models.PackageSource) *HTMLSourceResult {
+	var allVulns []HTMLVulnResult
+	var calledPackages = make(map[string]bool)
+	var uncalledPackages = make(map[string]bool)
+	var uncalledVulnIDs = make(map[string]bool)
+	var groupIDs = make(map[string]models.GroupInfo)
+	ecosystemName := ""
+
+	for _, vulnPkg := range packageSource.Packages {
+		if ecosystemName == "" {
+			ecosystemName = vulnPkg.Package.Ecosystem
+		}
+
+		// Process vulnerability groups and IDs to get called/uncalled information
+		processVulnerabilityGroups(groupIDs, vulnPkg, uncalledVulnIDs, calledPackages, uncalledPackages)
+		// Process vulnerabilities from one source package
+		allVulns = append(allVulns, processVulnerabilities(vulnPkg)...)
+	}
+
+	// Split vulnerabilities into called and uncalled.
+	// Only add one vulnerability per group
+	ecosystemPrefix := models.Ecosystem(strings.Split(ecosystemName, ":")[0])
+	packageResults := processPackageResults(allVulns, groupIDs, uncalledVulnIDs, ecosystemPrefix)
+	var count HTMLVulnCount
+	for index, packageResult := range packageResults {
+		packageResults[index].Ecosystem = ecosystemName
+		packageResults[index].Source = packageSource.Source.Path
+		updateCount(&count, &packageResult.HTMLVulnCount)
+	}
+
+	return &HTMLSourceResult{
+		Source:               packageSource.Source.String(),
+		Ecosystem:            ecosystemName,
+		PackageResults:       packageResults,
+		CalledPackageCount:   len(calledPackages),
+		UncalledPackageCount: len(uncalledPackages),
+		HTMLVulnCount:        count,
+	}
+}
+
+// processPackageResults splits the given vulnerabilities into called and uncalled
+// based on the uncalledVulnIDs map.
+func processPackageResults(allVulns []HTMLVulnResult, groupIDs map[string]models.GroupInfo, uncalledVulnIDs map[string]bool, ecosystemPrefix models.Ecosystem) []HTMLPackageResult {
+	packageResults := make(map[string]*HTMLPackageResult)
+	for _, vuln := range allVulns {
+		groupInfo, isIndex := groupIDs[vuln.Summary.ID]
+		if !isIndex {
+			// We only display one vulnerability from one group
+			continue
+		}
+
+		// Add group IDs info
+		if len(groupInfo.IDs) > 1 {
+			vuln.Detail.GroupIDs = strings.Join(groupInfo.IDs[1:], ", ")
```

Comment:
nit: in general it's better to leave all formatting decisions to the template itself, to enforce more separation between the logic (i.e. go) and the formatting (the template)

e.g. something similar to this example here: https://pkg.go.dev/text/template#example-Template-Block

```go
func main() {
	const (
		master  = `Names:{{block "list" .}}{{"\n"}}{{range .}}{{println "-" .}}{{end}}{{end}}`
		overlay = `{{define "list"}} {{join . ", "}}{{end}} `
	)
	var (
		funcs     = template.FuncMap{"join": strings.Join}
		guardians = []string{"Gamora", "Groot", "Nebula", "Rocket", "Star-Lord"}
	)
	masterTmpl, err := template.New("master").Funcs(funcs).Parse(master)
```

Is that doable here?

Reply:
Yeah, updated!